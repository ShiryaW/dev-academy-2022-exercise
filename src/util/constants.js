export const FARM_NAMES = {
  FRIMAN: "Friman Metsola Collective",
  PARTIALTECH: "PartialTech Research Farm",
  NOORA: "Noora's Farm",
  OSSI: "Organic Ossi's Impact That Lasts Plantation",
};

export const FARM_IDS = {
  [FARM_NAMES.FRIMAN]: 1,
  [FARM_NAMES.PARTIALTECH]: 2,
  [FARM_NAMES.NOORA]: 3,
  [FARM_NAMES.OSSI]: 4,
};
