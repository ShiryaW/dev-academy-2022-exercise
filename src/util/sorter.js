export const datesort = (date1, date2) => {
  const [day1, month1, year1] = parseDayMonthYear(date1);
  const [day2, month2, year2] = parseDayMonthYear(date2);

  if (year1 < year2) return -1;
  else if (year1 === year2) {
    if (month1 < month2) return -1;
    else if (month1 === month2) {
      if (day1 < day2) return -1;
      else if (day1 === day2) return 0;
      else return 1;
    } else return 1;
  } else return 1;
};

export const parseDayMonthYear = (date) => {
  return date
    .split(",")[0]
    .split(".")
    .map((n) => parseInt(n));
};
