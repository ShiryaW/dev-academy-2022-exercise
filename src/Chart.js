import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { VegaLite } from "react-vega";
import { ALL, FIELD_KEYS } from "./util/constants";

/* Example data entry:
  {
    datetime: "31.12.2018, 22:00:00",
    datetime_raw: "2018-12-31T22:00:00.000Z",
    farm_id: "2"
    id: 0
    location: "PartialTech Research Farm"
    sensor_type: "ph"
    value: 5.9
   }
 */

const specBase = {
  width: 1000,
  height: 300,
  data: { name: "measurements" },
  mark: "area",
  autosize: {
    resize: true,
  },
  encoding: {
    x: { field: FIELD_KEYS.DATETIME_RAW, type: "temporal", title: "Time" },
    y: { field: FIELD_KEYS.VALUE, type: "quantitative", title: "Temperature" },
  },
};

const specAllTypesAndFarms = {
  ...specBase,
  encoding: {
    ...specBase.encoding,
    color: { field: FIELD_KEYS.SENSOR_TYPE, type: "nominal", title: "Type" },
  },
};

export const Chart = ({ data, sensorType, selectedFarm }) => {
  const [spec, setSpec] = useState(specAllTypesAndFarms);

  useEffect(() => {
    getSpec(data.measurements, sensorType, selectedFarm);
  }, [data]);

  const getTypeFilter = (sensorType) => {
    return {
      transform: [
        {
          filter: {
            field: FIELD_KEYS.SENSOR_TYPE,
            equal: sensorType,
          },
        },
      ],
    };
  };

  const getSpec = (data, sensorType, selectedFarm) => {
    let spec = {
      ...specBase,
      ...getTypeFilter(sensorType),
    };

    if (selectedFarm === ALL) {
      spec.encoding = {
        ...specBase.encoding,
        color: {
          field: FIELD_KEYS.LOCATION,
          type: "nominal",
          title: "Location",
        },
      };
    }

    setSpec(spec);
  };

  return (
    <div id="chart">
      <VegaLite spec={spec} data={{ measurements: data.measurements }} />
    </div>
  );
};

Chart.propTypes = {
  data: PropTypes.object,
  sensorType: PropTypes.string,
  selectedFarm: PropTypes.string,
};
