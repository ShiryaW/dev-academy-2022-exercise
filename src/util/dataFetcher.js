import { FARM_IDS } from "./constants";

export const fetchFarmNames = async () => {
  const body = await fetch(`v1/farms`).then((res) => res.json());
  return body.map(({ farm_id, name }) => {
    return {
      id: farm_id,
      name,
    };
  });
};

export const getFarmStats = async (farm) => {
  return await fetch(`/v1/farms/${farm}/stats`).then((res) => res.json());
};

export const getStatsForAll = async () => {
  const promises = Object.values(FARM_IDS).map(
    async (id) =>
      await getFarmStats(id).then(({ measurements }) => measurements)
  );
  const measurements = await Promise.all(promises);
  return { measurements: measurements.flat() };
};
