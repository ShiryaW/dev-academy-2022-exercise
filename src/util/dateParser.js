import { ERRORS } from "./constants";

export const parseDate = (iso) => {
  // Input format: 2019-01-01T23:10:20.551Z
  if (!iso) throw new Error(ERRORS.MISSING_DATE);
  let date, time, dateArr, timeArr;

  try {
    [date, time] = iso.slice(0, -1).split("T");
    dateArr = date.split("-");
    timeArr = time.split(":");
  } catch (e) {
    throw new Error(ERRORS.MALFORMED_INPUT);
  }

  if (!date || !time || dateArr.length !== 3 || timeArr.length !== 3)
    throw new Error(ERRORS.MALFORMED_INPUT);

  return `${dateArr[2]}.${dateArr[1]}.${dateArr[0]}, ${timeArr[0]}:${
    timeArr[1]
  }:${Math.round(timeArr[2]).toString().padStart(2, "0")}`;
};

export const getDate = (iso) => {
  let result = ""; // fallback on an empty string if validation fails

  try {
    result = parseDate(iso);
  } catch (e) {
    if (e.message === ERRORS.MISSING_DATE) return e.message; // display placeholder text if there is no date
  }

  return result;
};
