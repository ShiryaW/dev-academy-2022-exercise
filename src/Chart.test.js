import React from "react";
import { shallow } from "enzyme";
import { Chart } from "./Chart";
import { VegaLite } from "react-vega";
import { ALL, SENSOR_TYPES } from "./util/constants";

const createChart = (props) =>
  shallow(
    <Chart
      {...{
        data: {
          measurements: [],
        },
        selectedFarm: ALL,
        ...props,
      }}
    />
  );

const testData = {
  measurements: [
    {
      location: "Friman Metsola collective",
      datetime: "2018-12-31T22:00:00.000Z",
      value: 6.52,
      farm_id: "1",
      sensor_type: "ph",
    },
    {
      location: "Friman Metsola collective",
      datetime: "2018-12-31T22:00:00.000Z",
      value: 2.6,
      farm_id: "1",
      sensor_type: "rainfall",
    },
    {
      location: "Friman Metsola collective",
      datetime: "2019-01-01T06:09:47.373Z",
      value: -9.0,
      farm_id: "1",
      sensor_type: "temperature",
    },
    {
      location: "Friman Metsola collective",
      datetime: "2019-01-01T21:04:09.813Z",
      value: -12.2,
      farm_id: "1",
      sensor_type: "temperature",
    },
  ],
};

describe("Chart", () => {
  it("renders the chart component empty when no data is supplied", () => {
    const wrapper = createChart();
    const chartInstance = wrapper.find(VegaLite);

    expect(chartInstance.exists()).toBe(true);
    expect(chartInstance.props()).toHaveProperty("data");
    expect(chartInstance.props().data).toHaveProperty("measurements");
    expect(chartInstance.props().data.measurements).toEqual([]);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the chart component with filtered data for temperature", () => {
    const wrapper = createChart({
      data: testData,
      sensorType: SENSOR_TYPES.TEMPERATURE,
    });
    const chartInstance = wrapper.find(VegaLite);

    expect(chartInstance.props().data.measurements).toEqual(
      testData.measurements
    );
    expect(chartInstance.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
});
