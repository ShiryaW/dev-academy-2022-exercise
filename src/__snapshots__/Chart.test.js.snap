// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart renders the chart component empty when no data is supplied 1`] = `
<div
  id="chart"
>
  <VegaLite
    data={
      Object {
        "measurements": Array [],
      }
    }
    spec={
      Object {
        "autosize": Object {
          "resize": true,
        },
        "data": Object {
          "name": "measurements",
        },
        "encoding": Object {
          "color": Object {
            "field": "sensor_type",
            "title": "Type",
            "type": "nominal",
          },
          "x": Object {
            "field": "datetime_raw",
            "title": "Time",
            "type": "temporal",
          },
          "y": Object {
            "field": "value",
            "title": "Temperature",
            "type": "quantitative",
          },
        },
        "height": 300,
        "mark": "area",
        "width": 1000,
      }
    }
  />
</div>
`;

exports[`Chart renders the chart component with filtered data for temperature 1`] = `
<div
  id="chart"
>
  <VegaLite
    data={
      Object {
        "measurements": Array [
          Object {
            "datetime": "2018-12-31T22:00:00.000Z",
            "farm_id": "1",
            "location": "Friman Metsola collective",
            "sensor_type": "ph",
            "value": 6.52,
          },
          Object {
            "datetime": "2018-12-31T22:00:00.000Z",
            "farm_id": "1",
            "location": "Friman Metsola collective",
            "sensor_type": "rainfall",
            "value": 2.6,
          },
          Object {
            "datetime": "2019-01-01T06:09:47.373Z",
            "farm_id": "1",
            "location": "Friman Metsola collective",
            "sensor_type": "temperature",
            "value": -9,
          },
          Object {
            "datetime": "2019-01-01T21:04:09.813Z",
            "farm_id": "1",
            "location": "Friman Metsola collective",
            "sensor_type": "temperature",
            "value": -12.2,
          },
        ],
      }
    }
    spec={
      Object {
        "autosize": Object {
          "resize": true,
        },
        "data": Object {
          "name": "measurements",
        },
        "encoding": Object {
          "color": Object {
            "field": "sensor_type",
            "title": "Type",
            "type": "nominal",
          },
          "x": Object {
            "field": "datetime_raw",
            "title": "Time",
            "type": "temporal",
          },
          "y": Object {
            "field": "value",
            "title": "Temperature",
            "type": "quantitative",
          },
        },
        "height": 300,
        "mark": "area",
        "width": 1000,
      }
    }
  />
</div>
`;
